File: algorithmic-toolbox/week2_algorithmic_warmup/4_least_common_multiple/lcm.py
# Uses python3
import random

def lcm_naive(a, b):
    for l in range(1, a*b + 1):
        if l % a == 0 and l % b == 0:
            return l

    return a*b

def gcd(a, b):
    if b == 0:
        return a

    rem = a % b
    return gcd(b, rem)

def lcm_fast(a, b):
    return int((a * b) / gcd(a, b))

def test_solution():
    assert lcm_fast(6, 8) == 24
    assert lcm_fast(761457, 614573) == 467970912861
    for _ in range(20):
        a, b = random.randint(1,100), random.randint(1, 100)
        assert lcm_fast(a, b) == lcm_naive(a, b)

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(lcm_fast(a, b))
    #print(lcm_naive(a, b))
    #test_solution()

File: algorithmic-toolbox/week2_algorithmic_warmup/4_least_common_multiple/lcm.cpp
#include <iostream>
#include <cassert>

long long lcm_naive(int a, int b) {
    for (long l = 1; l <= (long long) a * b; ++l)
        if (l % a == 0 && l % b == 0)
            return l;

    return (long long) a * b;
}

int gcd(int a, int b) {
    if (b == 0) {
        return a;
    }
    int rem = a % b;
    return gcd(b, rem);
}

long long lcm_fast(int a, int b) {
    return ((long long)a * b) / gcd(a, b);
}

void test_solution() {
    assert(lcm_fast(6, 8) == 24);
    assert(lcm_fast(761457, 614573) == 467970912861);
    for (int i = 1; i < 20; ++i) {
        int a = rand() % 100;
        int b = rand() % 100;
        assert(lcm_fast(a, b) == lcm_naive(a, b));
    }
}

int main() {
    int a, b;
    std::cin >> a >> b;
    //std::cout << lcm_naive(a, b) << std::endl;
    std::cout << lcm_fast(a, b) << std::endl;
    //test_solution();
    return 0;
}
